File: projects/src/main/java/io/nirahtech/erp/projects/utils/CriticalPathCalculator.java
package io.nirahtech.erp.projects.utils;

import java.util.Objects;
import java.util.OptionalLong;

import io.nirahtech.erp.projects.Project;
import io.nirahtech.erp.projects.Task;
import io.nirahtech.erp.projects.WorkPackage;

public final class CriticalPathCalculator {

    private CriticalPathCalculator() {
    }

    public static final long compute(final Project project) {
        final OptionalLong criticalPathInDays = project.getWorkPackages()
                .stream()
                .mapToLong(CriticalPathCalculator::compute)
                .max();
        return criticalPathInDays.orElse(0L);
    }

    public static final long compute(final WorkPackage workPackage) {
        final OptionalLong criticalPathInDays = workPackage.getTasks()
                .stream()
                .mapToLong(CriticalPathCalculator::compute)
                .max();
        return criticalPathInDays.orElse(0L);
    }

    public static final long compute(final Task task) {
        long days = 0;
        if (Objects.nonNull(task)) {
            if (Objects.isNull(task.getParent())) {
                days = task.getDuration().toDays();
            } else {
                days = task.getDuration().toDays() + CriticalPathCalculator.compute(task.getParent());
            }
        }
        return days;
    }
}

File: webapp/src/main/react/src/data/CriticalPathCalculator.ts
import Project from "./Project";
import { Task } from "./Task";
import { WorkPackage } from "./WorkPackage";

class CriticalPathCalculator {
  static computeProject(project: Project): number {
    const criticalPathInDays = Math.max(
      ...Array.from(project.getWorkPackages()).map((wp) => CriticalPathCalculator.computeWorkPackage(wp))
    );
    return criticalPathInDays;
  }

  static computeWorkPackage(workPackage: WorkPackage): number {
    const criticalPathInDays = Math.max(
      ...Array.from(workPackage.getTasks()).map((task) => CriticalPathCalculator.computeTask(task))
    );
    return criticalPathInDays;
  }

  static computeTask(task: Task): number {
    let days = 0;
    if (task !== null && task !== undefined) {
      let parent: Task|null = task.getParent();
      if (!parent) {
        days = task.getDuration().toDays();
      } else {
        days = task.getDuration().toDays() + CriticalPathCalculator.computeTask(parent);
      }
    }
    return days;
  }
}

export { CriticalPathCalculator };
